I have implemented parse_roman() function

AmPhIbIaN26 opened the pull request:
I have implemented the parse_roman() function.

Gallaecio commented:
Now I think we need to make `parse` and `parse_number` use this function as well.

I’m thinking that we may also want to include a new parameter to those functions, `numeral_systems`, which allows to limit the numeral systems to support while parsing. We could make it so that by default all numeral systems are used, but you can use `numeral_systems=['decimal']` to limit parsing to decimal numbers, or `numeral_systems=['roman']` to limit it to roman numbers.

For cases where users want to exclude a system, rather than include it, it may make sense to expose a public variable that contains the list of all supported numeral systems, e.g. `number_parser.NUMERAL_SYSTEMS`, so that users can create an exclusion-based subset (e.g. `[system for system in NUMERAL_SYSTEMS if system != 'roman']`).

Once these changes are done, I think we no longer need `parse_roman` itself, and it could be renamed as `_parse_roman` to discourage people from using it.

I’m also thinking that it may be possible to have slightly different implementations of `parse_roman` for each of the user-facing number-parser functions, for performance-tuning. But I haven’t look into it in detail, and what’s most important is that things work as intended, we can worry about performance later.

AmPhIbIaN26 commented:
I have added ```numeral_system``` as a parameter to functions which need it. I am going to add a ```_valid_input_by_numeral_system```  to check for numeral system if not given.

> For cases where users want to exclude a system, rather than include it, it may make sense to expose a public variable that contains the list of all supported numeral systems, e.g. `number_parser.NUMERAL_SYSTEMS`, so that users can create an exclusion-based subset (e.g. `[system for system in NUMERAL_SYSTEMS if system != 'roman']`).
> 

I still didn't understand how the user would use this  `[system for system in NUMERAL_SYSTEMS if system != 'roman']`

AmPhIbIaN26 commented:
Hi @Gallaecio hope you and your family are safe.

I added the  ```_valid_input_by_numeral_system()``` still have to improve this so it is easier to add another numeral system. I will work on that now and also improving the performance  of `_parse_roman()`.

AmPhIbIaN26 commented:
I have added `numeral_systems` as a parameter to `parse()`, the current system works like this
```python
all_numeral_systems_but_roman = [system for system in NUMERAL_SYSTEMS if system != 'roman']
all_numeral_systems_but_decimal = [system for system in NUMERAL_SYSTEMS if system != 'decimal']

>>>parse('Built in MMLXXVII.')
'Built in 2077.'

>>>parse( 'Built in MMLXXVII.', ['decimal'])
'Built in MMLXXVII.'

>>>parse('I was given two IV injections.', all_numeral_systems_but_roman)
'I was given 2 IV injections.'

>>>parse('I was given two IV injections.', all_numeral_systems_but_decimal)
1 was given two 4 injections.'

>>>parse('I was given two IV injections.')
'1 was given 2 4 injections.'

>>>parse('I have three apples.', all_numeral_systems_but_roman)
'I have 3 apples.'
```
I have added all these examples as test cases. I wanted to ask that since `parse_number()` only takes in 1 input how will having `numeral_systems` as a parameter help?

Gallaecio commented on number_parser/parser.py:
```diff
@@ -250,6 +251,9 @@ def parse_number(input_string, language=None):
     if input_string.strip().isnumeric():
         return int(input_string)
 
+    if re.search("^(m{0,3})(cm|cd|d?c{0,3})(xc|xl|l?x{0,3})(ix|iv|v?i{0,3})$", input_string.lower()):
```
Instead of lowercasing the input string, you can use `(?i)` at the very beginning of the pattern string to make it case-insensitive.

Gallaecio commented on number_parser/parser.py:
```diff
@@ -299,40 +303,33 @@ def parse_fraction(input_string, language=None):
     Converts all the numbers in a sentence written in natural language to their numeric type while keeping
     the other words unchanged. Returns the transformed string.
     """
+    global complete_sentence
```
Why is `global` used here?

Gallaecio commented on number_parser/parser.py:
```diff
@@ -299,40 +303,33 @@ def parse_fraction(input_string, language=None):
     """
+    global complete_sentence
+
+    if numeral_systems is None:
+        numeral_systems = ['decimal', 'roman']
```
You could just make `NUMERAL_SYSTEMS` the default value of the parameter in the function signature. Using lists as default parameters is often frown upon because if the list is changed the default parameter value changes as well, but this list is not meant to be edited. Also, to be sure, you could make it a tuple instead of a list.

Gallaecio commented on number_parser/parser.py:
```diff
@@ -299,40 +303,33 @@ def parse_fraction(input_string, language=None):
     if language is None:
         language = _valid_tokens_by_language(input_string)
 
+    for numeral_system in numeral_systems:
+
+        if numeral_system == 'decimal':
+            complete_sentence = _parse_decimal(input_string, language)
+            input_string = complete_sentence
+
+        if numeral_system == 'roman':
+            complete_sentence = _parse_roman(input_string)
+            input_string = complete_sentence
```
You could do `result = input_string` before the loop, and then stop using `input_string` (it can be confusing for it to contain something other than the actual input string) and simplify those pairs of calls to e.g. `result = _parse_roman(result)`.

You could also use `elif`, and you could include an `else` to raise an exception if an input numeral system is unknown.

Gallaecio commented on number_parser/parser.py:
```diff
@@ -386,8 +383,33 @@ def _build_and_add_number(pop_last_space=False):
+    tokens = _tokenize(input_string, None)
+    for token in tokens:
+        if re.search("^(m{0,3})(cm|cd|d?c{0,3})(xc|xl|l?x{0,3})(ix|iv|v?i{0,3})$", token.lower()):
+            if _build_roman(token) != 0:
+                tokens[tokens.index(token)] = str(_build_roman(token))
```
I would avoid calling `_build_roman` twice here.

Gallaecio commented on tests/data/permutations/all_roman_numbers.csv:
```diff
@@ -0,0 +1,3999 @@
+3995,MMMCMXCV
+3996,MMMCMXCVI
+3997,MMMCMXCVII
+3998,MMMCMXCVIII
```
I [read](https://blog.prepscholar.com/roman-numerals-converter#:~:text=*Though%20normally%20considered%20incorrect%20usage,numerals%20is%20MMMCMXCIX%2C%20or%203%2C999.) that `MMMCMXCIX` would be a valid `3999`.

Gallaecio commented on tests/data/permutations/all_roman_numbers.csv:
```diff
@@ -0,0 +1,3999 @@
+1,I
+2,II
+3,III
+4,IV
```
According to [this](https://blog.prepscholar.com/roman-numerals-converter#:~:text=*Though%20normally%20considered%20incorrect%20usage,numerals%20is%20MMMCMXCIX%2C%20or%203%2C999.), while technically incorrect, there’s cases were incorrect numbers like `IIII` instead of `IV` is used. Maybe we could support that as well.

Gallaecio commented:
> I have added all these examples as test cases. I wanted to ask that since parse_number() only takes in 1 input how will having numeral_systems as a parameter help?

It could allow users to fine-tune for performance if they know beforehand the numeral system of the input number, but making number-parser only use the desired number parser. Also, if they want parsing to simply fail for something like `I`, instead of returning 1.

AmPhIbIaN26 commented:
Thanks for looking into these, ill make the changes

AmPhIbIaN26 commented:
@Gallaecio hope you and your family are safe.
I have made all the changes you suggested also added support for incorrect roman numbers.

Gallaecio commented on number_parser/parser.py:
```diff
@@ -252,20 +259,37 @@ def parse_number(input_string, language=None):
+    if numeral_systems is None:
+        numeral_systems = NUMERAL_SYSTEMS
+
+    if _search_roman(input_string):
+        numeral_systems = ['roman']
```
`is_roman` may be better, since we are checking if the whole number is a valid Roman number, not if the string contains a Roman number. Something with a Roman number but non-Roman characters would not match here.

Also, this code ignores the input `numeral_systems` parameter. If it does not contain `roman`, this code should not do this, and instead should probably return `None` directly. This scenario is probably worth a test.

Similar (but opposite) with `decimal` below.

Gallaecio commented on number_parser/parser.py:
```diff
@@ -252,20 +259,37 @@ def parse_number(input_string, language=None):
+            return int(_parse_roman(input_string))
+
+        else:
+            raise ValueError(f'"{numeral_system}" is not a supported numeral system')
```
`!r` is a nice trick to get the value to be represented however it would be in Python (e.g. strings are quoted). It’s slightly better because if the input was of an invalid type (e.g. `numeral_systems=[b'roman']`) this would make it obvious, as it would print b'roman' instead of "roman" (or maybe there would be an exception because of the bytes needing encoding, come to think of it; not sure).

```suggestion
            raise ValueError(f'{numeral_system!r} is not a supported numeral system')
```

Gallaecio commented on number_parser/parser.py:
```diff
@@ -298,14 +322,37 @@ def parse_fraction(input_string, language=None):
     if language is None:
         language = _valid_tokens_by_language(input_string)
 
+    temporary_sentence = input_string
+    for numeral_system in numeral_systems:
+
+        if numeral_system == 'decimal':
+            complete_sentence = _parse_decimal(temporary_sentence, language)
+            temporary_sentence = complete_sentence
+
+        elif numeral_system == 'roman':
+            complete_sentence = _parse_roman(temporary_sentence)
+            temporary_sentence = complete_sentence
+
+        else:
+            raise ValueError(f'"{numeral_system}" is not a supported numeral system')
+
+    return complete_sentence
```
The new names still make you use 2 lines for something that could be expressed in 1 and remove 1 variable. In this case I think the simplification improves readability slightly.

```suggestion
    result = input_string
    for numeral_system in numeral_systems:

        if numeral_system == 'decimal':
            result = _parse_decimal(result, language)

        elif numeral_system == 'roman':
            result = _parse_roman(result)

        else:
            raise ValueError(f'"{numeral_system}" is not a supported numeral system')

    return result
```

AmPhIbIaN26 commented:
Hi @Gallaecio  hope you and your family are safe.

Any thoughts on this?

Gallaecio commented on number_parser/parser.py:
```diff
@@ -252,20 +259,35 @@ def parse_number(input_string, language=None):
+                    continue
+                if _is_skip_token(token, lang_data) and index != 0:
+                    continue
+                return None
+            number_built = _build_number(normalized_tokens, lang_data)
+            if len(number_built) == 1:
+                return int(number_built[0])
+            return None
```
Those `return None` seem like an issue now within a `for` loop. It looks like they will prevent the `for` loop to reach the next iteration (for the next numeral system).

Maybe you could move this code into a `_parse_decimal` function, and only return if the result is not `None`, else let the `for` loop go to the next numeral system.
